### ISSUE TITLE
$.parseHTML() creates anchor elements with empty href property in 1.12/2.2 in Chrome

### ISSUE THREAD

**Issue by vmagik** [jquery/jquery#2941]
$.parseHTML() seems to be broken in jQuery v1.12.0 and later for Chrome
This code shows empty string in Chrome but it is "test.html" in other browsers.
In jQuery 1.11.3 Chrome was OK too.
```html
<script type="text/javascript" src="jquery-1.12.1.js"></script>
<script type="text/javascript">
	function test()
	{
		var html = '<a href="test.html"></a>';
		var ref = $.parseHTML(html);

		alert(ref[0].href);
	}
</script>
```

EDIT (@mgol): I modified the `alert` to alert the `href` property of the anchor element; originally the test case was alerting the element (`alert(ref)`) which meant it relied on its stringification.

**Comment by dmethvin** [comment]
The result of `jQuery.parseHTML` should be some DOM, in this case an `<a>` element. It seems to be correct.

http://jsbin.com/balozatego/1/edit?js,console

 Do you have code that depends on stringifying DOM elements as you do in the example?

**Comment by timmywil** [comment]
Closing as worksforme. Please provide a test case to reopen.

**Comment by vmagik** [comment]
I checked the example above in Firefox 44, IE 11 and Chrome 48
Firefox and IE show "http://null.jsbin.com/test.html" in the first line of output,  Chrome - empty string 
If to use jQuery 1.11.3 all browsers show "http://null.jsbin.com/test.html"
Why the difference?

**Comment by mgol** [comment]
@vmagik Thanks for the update, we had an impression this report is about stringifying DOM elements but it seems the href property is, indeed, set to an empty string in Chrome. We need to investigate further.

**Comment by mgol** [comment]
OK, I nailed down the issue to commit 58c24608210c9a9a264a38746628ebc26823f59b. Here's a non-jQuery test case:

```js
var doc = document.implementation.createHTMLDocument();
var div = doc.createElement('div')
div.innerHTML = '<a href="foo.html"></a>';
document.body.innerHTML = 'URL: ' + div.childNodes[0].href;
```
Also posted at https://jsfiddle.net/ebacz2u5/2/. This test case fails in Chrome 48 and succeeds in Firefox 44, Safari 9.0, Edge 13, IE 9-11 (all that support `document.implementation.createHTMLDocument`), Opera 12.16 and even Android 2.3.

I see it reported upstream at https://bugs.chromium.org/p/chromium/issues/detail?id=568886

**Comment by dmethvin** [comment]
@vmagik How are you using this code in your real application? Are you calling `$.parseHTML` yourself or is something else calling it for you? We're trying to determine the impact of this.

**Comment by vmagik** [comment]
I call $.parseHTML() myself. 
My code gets a string like "<a ...</a>" and tries to get href from it. I didn't invent the code in the first message (with stringfication). Found an example somewhere.

**Comment by dmethvin** [comment]
What are you doing with the result though? Again we're trying to determine the seriousness of the bug based on the way things are being used in real code.

**Comment by vmagik** [comment]
I open a window to show resulting href. Also can use it for ajax call

**Comment by dmethvin** [comment]
You mean to show it to the user? I'm trying to get a sense of the work flow that is occurring here. Some type of HTML is parsed, where does it come from? Why is the href shown in a window?

**Comment by vmagik** [comment]
I show it to the user.
On click on some link on the page I first do ajax request which can result either in one link or several links.
If it's one link I just it in a new window or a dialog. If several links then I need to show user a choice.

**Comment by timmywil** [comment]
I guess if there's no workaround, we'll need to back out `document.implementation` in master too.

**Comment by timmywil** [comment]
Start with removal in 1.12.2/2.2.2. We're going to wait on removing in 3.0.

**Comment by gibson042** [comment]
This seems to be fixable by something [as simple as](https://jsfiddle.net/ebacz2u5/3/)
```js
var base = createdDoc.head.appendChild( createdDoc.createElement( "base" ) );
base.href = document.location.href;
```

Dirty, but not the first time we've had to spackle over browser behavior.

**Comment by timmywil** [comment]
@gibson042 Nice!

**Comment by dmethvin** [comment]
Since the caller can pass in a `context` of any document, does it work to set the base to an empty string? That way the href property/attribute would be equal.

I still would suggest we back this out in 1.x/2.x because we aren't using it consistently and it's not documented.

**Comment by mgol** [comment]
I like where @gibson042's going and I'd like to experiment with it on `master` but at this point all such changes seem risky to me for 1.12/2.2 which we're trying to stop being developed soon so I agree with @dmethvin we should back it out for 1.12.2/2.2.2.

**Comment by Elusive138** [comment]
Just saw this from the Chromium bugtracker. Interesting that it's affecting a larger library.

@vmagik - are you sure you're getting the resolved URL in Firefox? Your JSBin link shows just "test.html" in Firefox 44 and 46, the full "http://null.jsbin.com/test.html" in IE11 and "" in Chrome 48 for me.

@gibson042's fix looks like the correct way to go about it. Based on the [brief look I had at the specs](https://bugs.chromium.org/p/chromium/issues/detail?id=568886#c5), the default base URL is `about:blank`, so if your intention is to make the behaviour match the [`document` fallback](https://github.com/jquery/jquery/blob/58c24608210c9a9a264a38746628ebc26823f59b/src/core/parseHTML.js#L23) then you'd need to set the base URL anyway. Even if the Chromium bug is fixed correctly, you'd *still* need to do this to match `document` behaviour.

**Comment by mgol** [comment]
The backout has landed in c5c30735311c74c60689fcabdcf2cb192524000e on `2.2-stable` & 6403cf614fc2db42517005e80eef7bc5b2223d01 for `1.12-stable`. This issue is now resolved.

I created a separate one for making it work in 3.0: #2965.

**Comment by vmagik** [comment]
@Elusive138 Yes, Firefox 44 shows just "test.html" with jQuery 1.12.x. But with jQuery 1.11.3 it shows resolved URL.
http://jsbin.com/zuxevusoxi/edit?js,console
But I think this is not a big problem if a problem at all.
